lms/djangoapps/course_api/blocks/views.py

```diff
@@ -226,11 +227,17 @@ def list(self, request, usage_key_string, hide_access_denials=False):  # pylint:
                     hide_access_denials=hide_access_denials,
                 )
             )
+            # If the username is an empty string, and not None, then we are requesting
+            # data about the anonymous view of a course, which can be cached. In this
+            # case we add the usual caching headers to the response.
+            if params.cleaned_data.get('username', None) == '':
+                patch_response_headers(response)
```

Comment:
For my paranoia, can you also explicitly check to make sure we don't cache when `all_blocks=True`, even if the caller also does `username=` in the URL?

Reply:
@ormsbee  When `all_blocks=True` the code explicitly sets username to None, so this code would not trigger in that case.

If you use `all_blocks` as anything other than a staff user you will get a 403. If you use it along with `username`, the username will be set to None, so this code will never trigger. 

Relevant code: https://github.com/edx/edx-platform/pull/24264/files#diff-fc18b4a3c351e9d906f21cc6def6610eR117-R118

That said, to avoid any future I've added a test to specifically test for the situation where both `username=''` and `all_blocks=True`. It tests the response as a regular enrolled user, an unenrolled user, an anonymous user and staff user. In the first three it asserts that the response is 403 and not cached, and for staff, that the response is correct and not cached.